components/console/esp_linenoise.c

```diff
@@ -0,0 +1,1894 @@
+
+        case CTRL_K: /* Ctrl+k, delete from current to end of line. */
+
+            buf[l.pos] = '\0';
+            l.len = l.pos;
+            refreshLine(handle, &l);
+            break;
+
+        case CTRL_A: /* Ctrl+a, go to the start of the line */
+
+            editMoveHome(handle, &l);
+            break;
+
+        case CTRL_E: /* ctrl+e, go to the end of the line */
+
+            editMoveEnd(handle, &l);
+            break;
+
+        case CTRL_L: /* ctrl+l, clear screen */
+
+            esp_linenoise_clear_screen(handle);
+            refreshLine(handle, &l);
+            break;
+
+        case CTRL_W: /* ctrl+w, delete previous word */
+
+            editDeletePrevWord(handle, &l);
+            break;
+
+        }
+        flushWrite(handle);
+    }
+    return l.len;
+}
+
+static int linenoiseRaw(esp_linenoise_handle_t handle, char *buf, size_t buflen, const char *prompt)
+{
+    int count;
+
+    linenoiseOptions_t* opts = getOptions(handle);
+
+    if (buflen == 0) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    count = linenoiseEdit(handle, buf, buflen, prompt);
+    if (count != -1) {
+        fputc('\n', opts->stdout_fp);
+        flushWrite(handle);
+    }
+
+    return count;
+}
+
+static int linenoiseDumb(esp_linenoise_handle_t handle, char* buf, size_t buflen, const char* prompt)
+{
+    linenoiseOptions_t* opts = getOptions(handle);
+
+    /* dumb terminal, fall back to fgets */
+    fputs(prompt, opts->stdout_fp);
+    size_t count = 0;
+    while (count < buflen) {
+
+        int c = fgetc(opts->stdin_fp);
+        if (c == -1 && errno == EWOULDBLOCK) {
+            return -1;
+        }
+
+        if (c == '\n') {
+
+            break;
+
+        } else if (c >= 0x1c && c <= 0x1f) {
+
+            continue; /* consume arrow keys */
+
+        } else if (c == BACKSPACE || c == 0x8) {
+
+            if (count > 0) {
+                buf[count - 1] = 0;
+                count--;
+            }
+
+            /* Windows CMD: erase symbol under cursor */
+            fputs("\x08 ", opts->stdout_fp); 
+
+        } else {
+
+            buf[count] = c;
+            ++count;
+        }
+        fputc(c, opts->stdout_fp); /* echo */
+    }
+    fputc('\n', opts->stdout_fp);
+    flushWrite(handle);
+    return count;
+}
+
+static void sanitize(char* src) {
+    char* dst = src;
+    for (int c = *src; c != 0; src++, c = *src) {
+        if (isprint(c)) {
+            *dst = c;
+            ++dst;
+        }
+    }
+    *dst = 0;
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+//
+//
+// Public Api
+//
+//
+
+esp_linenoise_handle_t esp_linenoise_create()
+{   
+    struct esp_linenoise_t def = LINENOISE_DEFAULT;
+    esp_linenoise_handle_t handle = (esp_linenoise_handle_t) malloc(sizeof(struct esp_linenoise_t));
+    memcpy(handle, &def, sizeof(struct esp_linenoise_t));
+    return handle;
+}
+
+char *esp_linenoise(esp_linenoise_handle_t handle, const char *prompt)
```

Comment:
```suggestion
char *esp_linenoise_get_line(esp_linenoise_handle_t handle, const char *prompt)
```
or something like this — we don't have to keep the names 1:1 with the original version, and can make them more self-explanatory.